problem/0000-metrics-v1.md

```diff
@@ -0,0 +1,136 @@
+  - Peer Identity
+  - Peer K8s Labels? How?
+- Close Count, by the above as well as an errno.
+- Connect Latency Histogram, by the above
+- Connection Age Histogram, by the above
+- Bytes Transferred, by the above
+
+#### HTTP
+
+- Request count, by
+  - Interface: `inbound | outbound`
+  - Authority
+  - Workload
+  - Service
+  - Service profile route
+  - Peer Identity
+  - Peer K8s Labels? How?
+- Response count, by
+  - The above, plus
+  - Status Code; [100, 599]
+  - Classification: success | failure
+- Response Latency Histogram
+  - Should users be able to query at arbitrary percentiles?
+
+##### gRPC
+
+How do we represent the gRPC-specific metadata that Linkerd exposes? It seems like it should be
+distinct from (but in addition to) the HTTP-level abstractions? Or should HTTP metrics have a
+hole to punch through arbitrary metadata?
+
+Should we provide gRPC stats scoped by gRPC service name? We can infer these via known path
+structure when the appropriate headers are present.
+
+#### K8s services
+
+One of the main complexities in our metrics comes from handling Service-specific metadata.
+Currently all service-specific metadata is measured on the client-side, even when the querier
+asks about server-side resources. It is effectively impossible to query for server-side metrics
+for a service, since we allow pods to exist in more than one service. This is
+[fixable][rfc-src-metrics], however, if servers are able to discover metadata for inbound
+traffic.
+
+It seems like we should not support querying service metrics as a resource scope, though, because
+we cannot measure traffic _on_ a Service. We can measure outbound traffic from a set of client
+pods to a Service; and we can measure inbound traffic to a set of server pods that is targetted
+at a Service. I think our API should be explicit about these constraints.
+
+[rfc-src-metrics]: https://github.com/linkerd/rfc/pull/15
+
+### Prior Art
+
+#### Linkerd Stat
+
+Linkerd's [`Stat`][stat] API is resource oriented, but it only exposes an _outbound_ side, so
+there's no way to
+
+Why is `skip_stats` part of the API?
+
+The `tcp_stats` flag is a legacy of the CLI-specific nature of the command... it seems like we
+should either support a list of protocols or have protocol-specific endpoints.
+
+[l5d-stat]: https://github.com/linkerd/linkerd2/blob/dacf87e084a55757aab9cfa8556053495e43e207/proto/public.proto#L348-L427
+
+#### SMI
+
+[SMI's metrics][smi-spec] appear to be stringly typed, in that their are a few well-defined
+(http-specific?) metrics, but the base set of metrics is insufficient to satisfy Linkerd's CLI
+needs.
+
+However, the way it's resource-oriented is very appealing.
+
+It does not yet expose any time-series functionality.
+
+[smi-spec]: https://github.com/servicemeshinterface/smi-spec/blob/d5dd526a1a19c784065dc3fbac05a52decc01735/traffic-metrics.md
+
+#### Prometheus
+
+Can we expose a [Prometheus API][prom-api] scoped under each resource? E.g.
```

Comment:
This is very appealing and plays very nicely with RBAC.

Reply:
The more I think about this, the more I like it.  It also makes the transition easier for clients who are currently scraping Prometheus directly.

Reply:
I've got a nagging suspicion that this isn't going to work out all that well. Or that it will be wildly complex.

What happens if we have something like `../ns/linkerd/query`? How does RBAC apply? Must I have permission to read metrics for each pod/resource in the namespace? Or do I require an explicit RBAC binding on the namespace itself? Am I able to write a query that spans namespaces? How would that work?

What if we don't try to enforce RBAC on prometheus's _data_ at all?

A preferable model might be:

* Prometheus scrapes themselves are RBAC'd. I.e., the proxy applies access control to its own /metrics endpoints.
* The prometheus instance is the unit of view isolation. A user who can read from a prometheus instance has permission to read all of its data. If multi-tenancy is needed, users should deploy a Prometheus instance fore each tenant. RBAC is used to grant each instance permission to read metrics from the appropriate resources.
* Linkerd's challenge is then:
  * RBAC enforcement (for /metrics & prometheus access)
  * When a user interacts with Linkerd UIs how do we use the prometheus instance the user has access to?
    * Perhaps we make it easy to deploy a subset of the control plane per tenant? There's no good reason the `web` instance needs to be centralized in this world.
  * We publish a prometheus "schema" and scrape configs...

Basically: "Bring your own prometheus" should support arbitrary numbers of prometheus instances. Our users who really truly want to dump all of their data into one huge prometheus instance don't get view isolation (except by limiting who can read data from their monoprom).